packages/blocks/shuffle/src/shuffle.tsx

```diff
@@ -0,0 +1,88 @@
+import React, { useEffect, useReducer, useRef } from "react";
+import {
+  BlockComponent,
+  useGraphBlockService,
+} from "@blockprotocol/graph/react";
+// eslint-disable-next-line no-restricted-imports
+import Button from "@mui/material/Button";
+import Box from "@mui/material/Box";
+import { ItemList } from "./components/item-list";
+import { ActionType, initialItems, Items, shuffleReducer } from "./reducer";
+
+type BlockEntityProperties = {
+  items: Items;
+};
+
+export const Shuffle: BlockComponent<BlockEntityProperties> = ({
+  graph: {
+    blockEntity: {
+      entityId,
+      properties: { items },
+    },
+  },
+}) => {
+  const blockRootRef = useRef<HTMLDivElement>(null);
+  const { graphService } = useGraphBlockService(blockRootRef);
+
+  const [list, dispatch] = useReducer(
+    shuffleReducer,
+    items?.length ? items : initialItems,
+  );
+
+  const onReorder = (sourceIndex: number, destinationIndex: number) =>
+    dispatch({
+      type: ActionType.REORDER,
+      payload: {
+        sourceIndex,
+        destinationIndex,
+      },
+    });
+
+  const onValueChange = (index: number, value: string) =>
+    dispatch({
+      type: ActionType.UPDATE,
+      payload: {
+        index,
+        value,
+      },
+    });
+
+  const onAdd = (index: number) =>
+    dispatch({
+      type: ActionType.ADD,
+      payload: { index },
+    });
+
+  const onDelete = (index: number) =>
+    dispatch({
+      type: ActionType.DELETE,
+      payload: { index },
+    });
+
+  useEffect(() => {
+    void graphService?.updateEntity({
+      data: {
+        entityId,
+        properties: { items: list },
+      },
+    });
+  }, [graphService, entityId, list]);
```

Comment:
This effect will fire on the first render and when the `graphService` gets defined, which means two unnecessary calls when the block first loads.

One simple way to avoid this would be to check if `list` is different to `items`, and only call `updateEntity` if it is. 

Even better would be to call `updateEntity` when we know we want to call it, i.e. when the user adds, deletes, or shuffles the list. What this effect is trying to do is to call `updateEntity` when `list`  changes – but we already know when list changes, in response to user actions. So we should just call it when the user action happens directly, then rather than using this indirection.

cc @nathggns who explains this well and might have more thoughts

Reply:
I think you've explained this pretty perfectly. `useEffect` is bad for responding to user events, because all you know is something (maybe) changed, and not really what. Respond to events *in* event handlers. This kinda explains it https://beta.reactjs.org/learn/you-might-not-need-an-effect#sharing-logic-between-event-handlers

Reply:
Thank you both. Addressed in 1151c80b95b319e555d33f85a69adb9a2c8f57c5

I am now updating both the state and the entity in the db in the event handlers. I couldn't find a clean way to do this with useReducer so I dropped the reducer and went with useState as most of the other blocks have done.